[Rustbuild] Found possibly newer version of crate `std` which `rustc_serialize` depends on

It's hard to tell exactly what causes this sort of error, but I see similar errors quite often after pulling from GitHub. Let me know if there's other artifacts or anything else I can provide to help diagnose this.

Verbose log:
```
$ ./x.py test src/test/run-pass --stage 1 --verbose
    Finished debug [unoptimized] target(s) in 0.0 secs
finding compilers
CC_x86_64-unknown-linux-gnu = "cc"
AR_x86_64-unknown-linux-gnu = "ar"
CXX_x86_64-unknown-linux-gnu = "c++"
running sanity check
collecting channel variables
updating submodules
running: "git" "submodule" "sync"
Synchronizing submodule url for 'src/compiler-rt'
Synchronizing submodule url for 'src/jemalloc'
Synchronizing submodule url for 'src/liblibc'
Synchronizing submodule url for 'src/llvm'
Synchronizing submodule url for 'src/rt/hoedown'
Synchronizing submodule url for 'src/rust-installer'
running: "git" "reset" "--hard"
HEAD is now at a8fc4c16 Merge pull request #28 from xen0n/preprocessor-firefighting
running: "git" "clean" "-fdx"
running: "git" "reset" "--hard"
HEAD is now at 11bfb0d Merge pull request #16 from glandium/rust
running: "git" "clean" "-fdx"
running: "git" "reset" "--hard"
HEAD is now at 7d57bdcd Auto merge of #484 - eddyb:dox-unops, r=alexcrichton
running: "git" "clean" "-fdx"
running: "git" "reset" "--hard"
HEAD is now at ceb177eeefa Merge pull request #60 from japaric/gh38406
running: "git" "clean" "-fdx"
running: "git" "reset" "--hard"
HEAD is now at a3736a0 Merge pull request #6 from intelfx/patch-1
running: "git" "clean" "-fdx"
running: "git" "reset" "--hard"
HEAD is now at 4f99485 Merge pull request #54 from brson/docdir
running: "git" "clean" "-fdx"
learning about cargo
bootstrap top targets:
	Step { name: "check-rpass", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
bootstrap build plan:
	Step { name: "create-sysroot", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "create-sysroot", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "rustc", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "startup-objects", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "build-crate-std_shim", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "libstd-link", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "libstd", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "build-crate-test_shim", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "libtest-link", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "libtest", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "llvm", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "build-crate-rustc-main", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "librustc-link", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "librustc", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "rustc", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "startup-objects", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "build-crate-std_shim", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "libstd-link", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "libstd", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "build-crate-test_shim", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "libtest-link", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "libtest", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "tool-compiletest", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "test-helpers", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "android-copy-libs", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
	Step { name: "check-rpass", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "create-sysroot", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "create-sysroot", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "rustc", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "startup-objects", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "build-crate-std_shim", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
running: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "-j" "8" "--target" "x86_64-unknown-linux-gnu" "-v" "--release" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/home/mark/Build/rust/src/rustc/std_shim/Cargo.toml"
       Fresh build_helper v0.1.0 (file:///home/mark/Build/rust/src/build_helper)
       Fresh core v0.0.0 (file:///home/mark/Build/rust/src/libcore)
       Fresh gcc v0.3.40
       Fresh alloc v0.0.0 (file:///home/mark/Build/rust/src/liballoc)
       Fresh std_unicode v0.0.0 (file:///home/mark/Build/rust/src/libstd_unicode)
       Fresh rand v0.0.0 (file:///home/mark/Build/rust/src/librand)
       Fresh collections v0.0.0 (file:///home/mark/Build/rust/src/libcollections)
       Fresh libc v0.0.0 (file:///home/mark/Build/rust/src/rustc/libc_shim)
       Fresh compiler_builtins v0.0.0 (file:///home/mark/Build/rust/src/libcompiler_builtins)
       Fresh panic_abort v0.0.0 (file:///home/mark/Build/rust/src/libpanic_abort)
       Fresh unwind v0.0.0 (file:///home/mark/Build/rust/src/libunwind)
       Fresh alloc_system v0.0.0 (file:///home/mark/Build/rust/src/liballoc_system)
       Fresh panic_unwind v0.0.0 (file:///home/mark/Build/rust/src/libpanic_unwind)
       Fresh alloc_jemalloc v0.0.0 (file:///home/mark/Build/rust/src/liballoc_jemalloc)
       Fresh std v0.0.0 (file:///home/mark/Build/rust/src/libstd)
       Fresh std_shim v0.1.0 (file:///home/mark/Build/rust/src/rustc/std_shim)
    Finished release [optimized] target(s) in 0.0 secs
executing step Step { name: "libstd-link", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Copying stage0 std from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
executing step Step { name: "libstd", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "build-crate-test_shim", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Building stage0 test artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
running: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "-j" "8" "--target" "x86_64-unknown-linux-gnu" "-v" "--release" "--manifest-path" "/home/mark/Build/rust/src/rustc/test_shim/Cargo.toml"
       Fresh term v0.0.0 (file:///home/mark/Build/rust/src/libterm)
       Fresh getopts v0.0.0 (file:///home/mark/Build/rust/src/libgetopts)
       Fresh test v0.0.0 (file:///home/mark/Build/rust/src/libtest)
       Fresh test_shim v0.1.0 (file:///home/mark/Build/rust/src/rustc/test_shim)
    Finished release [optimized] target(s) in 0.0 secs
executing step Step { name: "libtest-link", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Copying stage0 test from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
executing step Step { name: "libtest", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "llvm", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "build-crate-rustc-main", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Building stage0 compiler artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
running: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "-j" "8" "--target" "x86_64-unknown-linux-gnu" "-v" "--release" "--features" " jemalloc" "--manifest-path" "/home/mark/Build/rust/src/rustc/Cargo.toml"
       Fresh rustc_bitflags v0.0.0 (file:///home/mark/Build/rust/src/librustc_bitflags)
       Fresh log v0.0.0 (file:///home/mark/Build/rust/src/liblog)
       Fresh gcc v0.3.40
       Fresh arena v0.0.0 (file:///home/mark/Build/rust/src/libarena)
       Fresh fmt_macros v0.0.0 (file:///home/mark/Build/rust/src/libfmt_macros)
       Fresh graphviz v0.0.0 (file:///home/mark/Build/rust/src/libgraphviz)
       Fresh rustc_i128 v0.0.0 (file:///home/mark/Build/rust/src/librustc_i128)
       Fresh build_helper v0.1.0 (file:///home/mark/Build/rust/src/build_helper)
       Fresh rustc_platform_intrinsics v0.0.0 (file:///home/mark/Build/rust/src/librustc_platform_intrinsics)
       Fresh serialize v0.0.0 (file:///home/mark/Build/rust/src/libserialize)
       Fresh syntax_pos v0.0.0 (file:///home/mark/Build/rust/src/libsyntax_pos)
       Fresh rustc_data_structures v0.0.0 (file:///home/mark/Build/rust/src/librustc_data_structures)
       Fresh rustc_errors v0.0.0 (file:///home/mark/Build/rust/src/librustc_errors)
       Fresh flate v0.0.0 (file:///home/mark/Build/rust/src/libflate)
       Fresh rustc_llvm v0.0.0 (file:///home/mark/Build/rust/src/librustc_llvm)
       Fresh syntax v0.0.0 (file:///home/mark/Build/rust/src/libsyntax)
       Fresh proc_macro v0.0.0 (file:///home/mark/Build/rust/src/libproc_macro)
       Fresh rustc_const_math v0.0.0 (file:///home/mark/Build/rust/src/librustc_const_math)
       Fresh rustc_back v0.0.0 (file:///home/mark/Build/rust/src/librustc_back)
       Fresh syntax_ext v0.0.0 (file:///home/mark/Build/rust/src/libsyntax_ext)
       Fresh rustc v0.0.0 (file:///home/mark/Build/rust/src/librustc)
       Fresh rustc_privacy v0.0.0 (file:///home/mark/Build/rust/src/librustc_privacy)
       Fresh rustc_const_eval v0.0.0 (file:///home/mark/Build/rust/src/librustc_const_eval)
       Fresh rustc_metadata v0.0.0 (file:///home/mark/Build/rust/src/librustc_metadata)
       Fresh rustc_resolve v0.0.0 (file:///home/mark/Build/rust/src/librustc_resolve)
       Fresh rustc_save_analysis v0.0.0 (file:///home/mark/Build/rust/src/librustc_save_analysis)
       Fresh rustc_incremental v0.0.0 (file:///home/mark/Build/rust/src/librustc_incremental)
       Fresh rustc_typeck v0.0.0 (file:///home/mark/Build/rust/src/librustc_typeck)
       Fresh rustc_mir v0.0.0 (file:///home/mark/Build/rust/src/librustc_mir)
       Fresh rustc_passes v0.0.0 (file:///home/mark/Build/rust/src/librustc_passes)
       Fresh rustc_lint v0.0.0 (file:///home/mark/Build/rust/src/librustc_lint)
       Fresh rustc_plugin v0.0.0 (file:///home/mark/Build/rust/src/librustc_plugin)
       Fresh rustc_trans v0.0.0 (file:///home/mark/Build/rust/src/librustc_trans)
       Fresh rustc_borrowck v0.0.0 (file:///home/mark/Build/rust/src/librustc_borrowck)
       Fresh proc_macro_plugin v0.0.0 (file:///home/mark/Build/rust/src/libproc_macro_plugin)
       Fresh rustc_driver v0.0.0 (file:///home/mark/Build/rust/src/librustc_driver)
       Fresh rustdoc v0.0.0 (file:///home/mark/Build/rust/src/librustdoc)
       Fresh rustc-main v0.0.0 (file:///home/mark/Build/rust/src/rustc)
    Finished release [optimized] target(s) in 0.0 secs
executing step Step { name: "librustc-link", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Copying stage0 rustc from stage0 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
executing step Step { name: "librustc", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "rustc", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Copying stage1 compiler (x86_64-unknown-linux-gnu)
executing step Step { name: "startup-objects", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "build-crate-std_shim", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Building stage1 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
running: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "-j" "8" "--target" "x86_64-unknown-linux-gnu" "-v" "--release" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/home/mark/Build/rust/src/rustc/std_shim/Cargo.toml"
       Fresh core v0.0.0 (file:///home/mark/Build/rust/src/libcore)
       Fresh build_helper v0.1.0 (file:///home/mark/Build/rust/src/build_helper)
       Fresh gcc v0.3.40
       Fresh rand v0.0.0 (file:///home/mark/Build/rust/src/librand)
       Fresh alloc v0.0.0 (file:///home/mark/Build/rust/src/liballoc)
       Fresh std_unicode v0.0.0 (file:///home/mark/Build/rust/src/libstd_unicode)
       Fresh collections v0.0.0 (file:///home/mark/Build/rust/src/libcollections)
       Fresh libc v0.0.0 (file:///home/mark/Build/rust/src/rustc/libc_shim)
       Fresh unwind v0.0.0 (file:///home/mark/Build/rust/src/libunwind)
       Fresh panic_abort v0.0.0 (file:///home/mark/Build/rust/src/libpanic_abort)
       Fresh alloc_system v0.0.0 (file:///home/mark/Build/rust/src/liballoc_system)
       Fresh compiler_builtins v0.0.0 (file:///home/mark/Build/rust/src/libcompiler_builtins)
       Fresh alloc_jemalloc v0.0.0 (file:///home/mark/Build/rust/src/liballoc_jemalloc)
       Fresh panic_unwind v0.0.0 (file:///home/mark/Build/rust/src/libpanic_unwind)
       Fresh std v0.0.0 (file:///home/mark/Build/rust/src/libstd)
       Fresh std_shim v0.1.0 (file:///home/mark/Build/rust/src/rustc/std_shim)
    Finished release [optimized] target(s) in 0.0 secs
executing step Step { name: "libstd-link", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Copying stage1 std from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
executing step Step { name: "libstd", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "build-crate-test_shim", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Building stage1 test artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
running: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "-j" "8" "--target" "x86_64-unknown-linux-gnu" "-v" "--release" "--manifest-path" "/home/mark/Build/rust/src/rustc/test_shim/Cargo.toml"
       Fresh getopts v0.0.0 (file:///home/mark/Build/rust/src/libgetopts)
       Fresh term v0.0.0 (file:///home/mark/Build/rust/src/libterm)
       Fresh test v0.0.0 (file:///home/mark/Build/rust/src/libtest)
       Fresh test_shim v0.1.0 (file:///home/mark/Build/rust/src/rustc/test_shim)
    Finished release [optimized] target(s) in 0.0 secs
executing step Step { name: "libtest-link", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Copying stage1 test from stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu / x86_64-unknown-linux-gnu)
executing step Step { name: "libtest", stage: 1, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
executing step Step { name: "tool-compiletest", stage: 0, host: "x86_64-unknown-linux-gnu", target: "x86_64-unknown-linux-gnu" }
Building stage0 tool compiletest (x86_64-unknown-linux-gnu)
running: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "-j" "8" "--target" "x86_64-unknown-linux-gnu" "-v" "--release" "--manifest-path" "/home/mark/Build/rust/src/tools/compiletest/Cargo.toml"
       Fresh log v0.3.6
       Fresh rustc-serialize v0.3.19
       Fresh env_logger v0.3.5
   Compiling compiletest v0.0.0 (file:///home/mark/Build/rust/src/tools/compiletest)
     Running `/home/mark/Build/rust/build/bootstrap/debug/rustc --crate-name compiletest src/tools/compiletest/src/main.rs --crate-type bin -C opt-level=2 -C metadata=73c24ea7041cbcca -C extra-filename=-73c24ea7041cbcca --out-dir /home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps --emit=dep-info,link --target x86_64-unknown-linux-gnu -L dependency=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps -L dependency=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/release/deps --extern log=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/liblog-8fea599810f87f11.rlib --extern rustc_serialize=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/librustc_serialize-ae20735284b6aaa6.rlib --extern env_logger=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/libenv_logger-2ed6b4d5cbd2041d.rlib --cfg cargobuild`
error[E0460]: found possibly newer version of crate `std` which `rustc_serialize` depends on
  --> src/tools/compiletest/src/main.rs:26:1
   |
26 | extern crate rustc_serialize;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: perhaps that crate needs to be recompiled?
   = note: crate `std` path #1: /home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-6f143d936af1f194.rlib
   = note: crate `std` path #2: /home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-6f143d936af1f194.so
   = note: crate `rustc_serialize` path #1: /home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/librustc_serialize-ae20735284b6aaa6.rlib

error: aborting due to previous error

error: Could not compile `compiletest`.

Caused by:
  process didn't exit successfully: `/home/mark/Build/rust/build/bootstrap/debug/rustc --crate-name compiletest src/tools/compiletest/src/main.rs --crate-type bin -C opt-level=2 -C metadata=73c24ea7041cbcca -C extra-filename=-73c24ea7041cbcca --out-dir /home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps --emit=dep-info,link --target x86_64-unknown-linux-gnu -L dependency=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps -L dependency=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/release/deps --extern log=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/liblog-8fea599810f87f11.rlib --extern rustc_serialize=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/librustc_serialize-ae20735284b6aaa6.rlib --extern env_logger=/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/deps/libenv_logger-2ed6b4d5cbd2041d.rlib --cfg cargobuild` (exit code: 101)


command did not execute successfully: "/home/mark/Build/rust/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "-j" "8" "--target" "x86_64-unknown-linux-gnu" "-v" "--release" "--manifest-path" "/home/mark/Build/rust/src/tools/compiletest/Cargo.toml"
expected success, got: exit code: 101
```

cc @alexcrichton

--- a/src/bootstrap/compile.rs
+++ b/src/bootstrap/compile.rs
@@ -275,6 +275,7 @@ pub fn rustc(build: &Build, target: &str, compiler: &Compiler) {
         cargo.env("CFG_DEFAULT_AR", s);
     }
     build.run(&mut cargo);
+    update_mtime(build, &librustc_stamp(build, compiler, target));
 }
 
 /// Same as `std_link`, only for librustc
@@ -303,6 +304,12 @@ fn libstd_stamp(build: &Build, compiler: &Compiler, target: &str) -> PathBuf {
 /// compiler for the specified target.
 fn libtest_stamp(build: &Build, compiler: &Compiler, target: &str) -> PathBuf {
     build.cargo_out(compiler, Mode::Libtest, target).join(".libtest.stamp")
+}
+
+/// Cargo's output path for librustc in a given stage, compiled by a particular
+/// compiler for the specified target.
+fn librustc_stamp(build: &Build, compiler: &Compiler, target: &str) -> PathBuf {
+    build.cargo_out(compiler, Mode::Librustc, target).join(".librustc.stamp")
 }
 
 fn compiler_file(compiler: &Path, file: &str) -> PathBuf {
@@ -411,19 +418,27 @@ fn add_to_sysroot(out_dir: &Path, sysroot_dst: &Path) {
 ///
 /// This will build the specified tool with the specified `host` compiler in
 /// `stage` into the normal cargo output directory.
+pub fn maybe_clean_tools(build: &Build, stage: u32, target: &str, mode: Mode) {
+    let compiler = Compiler::new(stage, &build.config.build);
+
+    let stamp = match mode {
+        Mode::Libstd => libstd_stamp(build, &compiler, target),
+        Mode::Libtest => libtest_stamp(build, &compiler, target),
+        Mode::Librustc => librustc_stamp(build, &compiler, target),
+        _ => panic!(),
+    };
+    let out_dir = build.cargo_out(&compiler, Mode::Tool, target);
+    build.clear_if_dirty(&out_dir, &stamp);
+}
+
+/// Build a tool in `src/tools`
+///
+/// This will build the specified tool with the specified `host` compiler in
+/// `stage` into the normal cargo output directory.
 pub fn tool(build: &Build, stage: u32, target: &str, tool: &str) {
     println!("Building stage{} tool {} ({})", stage, tool, target);
 
     let compiler = Compiler::new(stage, &build.config.build);
-
-    // FIXME: need to clear out previous tool and ideally deps, may require
-    //        isolating output directories or require a pseudo shim step to
-    //        clear out all the info.
-    //
-    //        Maybe when libstd is compiled it should clear out the rustc of the
-    //        corresponding stage?
-    // let out_dir = build.cargo_out(stage, &host, Mode::Librustc, target);
-    // build.clear_if_dirty(&out_dir, &libstd_stamp(build, stage, &host, target));
 
     let mut cargo = build.cargo(&compiler, Mode::Tool, target, "build");
     let mut dir = build.src.join(tool);

--- a/src/bootstrap/step.rs
+++ b/src/bootstrap/step.rs
@@ -26,7 +26,7 @@
 //! along with the actual implementation elsewhere. You can find more comments
 //! about how to define rules themselves below.
 
-use std::collections::{BTreeMap, HashSet};
+use std::collections::{BTreeMap, HashSet, HashMap};
 use std::mem;
 
 use check::{self, TestKind};
@@ -533,34 +533,44 @@ pub fn build_rules<'a>(build: &'a Build) -> Rules {
     //
     // Tools used during the build system but not shipped
     rules.build("tool-rustbook", "src/tools/rustbook")
-         .dep(|s| s.name("librustc"))
+         .dep(|s| s.name("maybe-clean-tools"))
+         .dep(|s| s.name("librustc-tool"))
          .run(move |s| compile::tool(build, s.stage, s.target, "rustbook"));
     rules.build("tool-error-index", "src/tools/error_index_generator")
-         .dep(|s| s.name("librustc"))
+         .dep(|s| s.name("maybe-clean-tools"))
+         .dep(|s| s.name("librustc-tool"))
          .run(move |s| compile::tool(build, s.stage, s.target, "error_index_generator"));
     rules.build("tool-tidy", "src/tools/tidy")
-         .dep(|s| s.name("libstd"))
+         .dep(|s| s.name("maybe-clean-tools"))
+         .dep(|s| s.name("libstd-tool"))
          .run(move |s| compile::tool(build, s.stage, s.target, "tidy"));
     rules.build("tool-linkchecker", "src/tools/linkchecker")
-         .dep(|s| s.name("libstd"))
+         .dep(|s| s.name("maybe-clean-tools"))
+         .dep(|s| s.name("libstd-tool"))
          .run(move |s| compile::tool(build, s.stage, s.target, "linkchecker"));
     rules.build("tool-cargotest", "src/tools/cargotest")
-         .dep(|s| s.name("libstd"))
+         .dep(|s| s.name("maybe-clean-tools"))
+         .dep(|s| s.name("libstd-tool"))
          .run(move |s| compile::tool(build, s.stage, s.target, "cargotest"));
     rules.build("tool-compiletest", "src/tools/compiletest")
-         .dep(|s| s.name("libtest"))
+         .dep(|s| s.name("maybe-clean-tools"))
+         .dep(|s| s.name("libtest-tool"))
          .run(move |s| compile::tool(build, s.stage, s.target, "compiletest"));
     rules.build("tool-build-manifest", "src/tools/build-manifest")
-         .dep(|s| s.name("libstd"))
+         .dep(|s| s.name("maybe-clean-tools"))
+         .dep(|s| s.name("libstd-tool"))
          .run(move |s| compile::tool(build, s.stage, s.target, "build-manifest"));
     rules.build("tool-qemu-test-server", "src/tools/qemu-test-server")
-         .dep(|s| s.name("libstd"))
+         .dep(|s| s.name("maybe-clean-tools"))
+         .dep(|s| s.name("libstd-tool"))
          .run(move |s| compile::tool(build, s.stage, s.target, "qemu-test-server"));
     rules.build("tool-qemu-test-client", "src/tools/qemu-test-client")
-         .dep(|s| s.name("libstd"))
+         .dep(|s| s.name("maybe-clean-tools"))
+         .dep(|s| s.name("libstd-tool"))
          .run(move |s| compile::tool(build, s.stage, s.target, "qemu-test-client"));
     rules.build("tool-cargo", "cargo")
-         .dep(|s| s.name("libstd"))
+         .dep(|s| s.name("maybe-clean-tools"))
+         .dep(|s| s.name("libstd-tool"))
          .dep(|s| s.stage(0).host(s.target).name("openssl"))
          .dep(move |s| {
              // Cargo depends on procedural macros, which requires a full host
@@ -572,7 +582,8 @@ pub fn build_rules<'a>(build: &'a Build) -> Rules {
          .run(move |s| compile::tool(build, s.stage, s.target, "cargo"));
     rules.build("tool-rls", "rls")
          .host(true)
-         .dep(|s| s.name("librustc"))
+         .dep(|s| s.name("librustc-tool"))
+         .dep(|s| s.stage(0).host(s.target).name("openssl"))
          .dep(move |s| {
              // rls, like cargo, uses procedural macros
              s.name("librustc-link")
@@ -580,6 +591,25 @@ pub fn build_rules<'a>(build: &'a Build) -> Rules {
               .host(&build.config.build)
          })
          .run(move |s| compile::tool(build, s.stage, s.target, "rls"));
+
+    // "pseudo rule" which represents completely cleaning out the tools dir in
+    // one stage. This needs to happen whenever a dependency changes (e.g.
+    // libstd, libtest, librustc) and all of the tool compilations above will
+    // be sequenced after this rule.
+    rules.build("maybe-clean-tools", "path/to/nowhere")
+         .after("librustc-tool")
+         .after("libtest-tool")
+         .after("libstd-tool");
+
+    rules.build("librustc-tool", "path/to/nowhere")
+         .dep(|s| s.name("librustc"))
+         .run(move |s| compile::maybe_clean_tools(build, s.stage, s.target, Mode::Librustc));
+    rules.build("libtest-tool", "path/to/nowhere")
+         .dep(|s| s.name("libtest"))
+         .run(move |s| compile::maybe_clean_tools(build, s.stage, s.target, Mode::Libtest));
+    rules.build("libstd-tool", "path/to/nowhere")
+         .dep(|s| s.name("libstd"))
+         .run(move |s| compile::maybe_clean_tools(build, s.stage, s.target, Mode::Libstd));
 
     // ========================================================================
     // Documentation targets
@@ -828,6 +858,11 @@ struct Rule<'a> {
     /// Whether this rule is only for the build triple, not anything in hosts or
     /// targets.
     only_build: bool,
+
+    /// A list of "order only" dependencies. This rules does not actually
+    /// depend on these rules, but if they show up in the dependency graph then
+    /// this rule must be executed after all these rules.
+    after: Vec<&'a str>,
 }
 
 #[derive(PartialEq)]
@@ -851,6 +886,7 @@ impl<'a> Rule<'a> {
             host: false,
             only_host_build: false,
             only_build: false,
+            after: Vec::new(),
         }
     }
 }
@@ -867,6 +903,11 @@ impl<'a, 'b> RuleBuilder<'a, 'b> {
         where F: Fn(&Step<'a>) -> Step<'a> + 'a,
     {
         self.rule.deps.push(Box::new(f));
+        self
+    }
+
+    fn after(&mut self, step: &'a str) -> &mut Self {
+        self.rule.after.push(step);
         self
     }
 
@@ -1153,31 +1194,52 @@ invalid rule dependency graph detected, was a rule added and maybe typo'd?
     /// From the top level targets `steps` generate a topological ordering of
     /// all steps needed to run those steps.
     fn expand(&self, steps: &[Step<'a>]) -> Vec<Step<'a>> {
+        // First up build a graph of steps and their dependencies. The `nodes`
+        // map is a map from step to a unique number. The `edges` map is a
+        // map from these unique numbers to a list of other numbers,
+        // representing dependencies.
+        let mut nodes = HashMap::new();
+        nodes.insert(Step::noop(), 0);
+        let mut edges = HashMap::new();
+        edges.insert(0, HashSet::new());
+        for step in steps {
+            self.build_graph(step.clone(), &mut nodes, &mut edges);
+        }
+
+        // Now that we've built up the actual dependency graph, draw more
+        // dependency edges to satisfy the `after` dependencies field for each
+        // rule.
+        self.satisfy_after_deps(&nodes, &mut edges);
+
+        // And finally, perform a topological sort to return a list of steps to
+        // execute.
         let mut order = Vec::new();
-        let mut added = HashSet::new();
-        added.insert(Step::noop());
-        for step in steps.iter().cloned() {
-            self.fill(step, &mut order, &mut added);
+        let mut visited = HashSet::new();
+        visited.insert(0);
+        let idx_to_node = nodes.iter().map(|p| (*p.1, p.0)).collect::<HashMap<_, _>>();
+        for idx in nodes.values() {
+            self.topo_sort(*idx, &idx_to_node, &edges, &mut visited, &mut order);
         }
         return order
     }
 
-    /// Performs topological sort of dependencies rooted at the `step`
-    /// specified, pushing all results onto the `order` vector provided.
+    /// Builds the dependency graph rooted at `step`.
     ///
-    /// In other words, when this method returns, the `order` vector will
-    /// contain a list of steps which if executed in order will eventually
-    /// complete the `step` specified as well.
-    ///
-    /// The `added` set specified here is the set of steps that are already
-    /// present in `order` (and hence don't need to be added again).
-    fn fill(&self,
-            step: Step<'a>,
-            order: &mut Vec<Step<'a>>,
-            added: &mut HashSet<Step<'a>>) {
-        if !added.insert(step.clone()) {
-            return
+    /// The `nodes` and `edges` maps are filled out according to the rule
+    /// described by `step.name`.
+    fn build_graph(&self,
+                   step: Step<'a>,
+                   nodes: &mut HashMap<Step<'a>, usize>,
+                   edges: &mut HashMap<usize, HashSet<usize>>) -> usize {
+        use std::collections::hash_map::Entry;
+
+        let idx = nodes.len();
+        match nodes.entry(step.clone()) {
+            Entry::Vacant(e) => { e.insert(idx); }
+            Entry::Occupied(e) => return *e.get(),
         }
+
+        let mut deps = Vec::new();
         for dep in self.rules[step.name].deps.iter() {
             let dep = dep(&step);
             if dep.name.starts_with("default:") {
@@ -1189,13 +1251,61 @@ invalid rule dependency graph detected, was a rule added and maybe typo'd?
                 let host = self.build.config.host.iter().any(|h| h == dep.target);
                 let rules = self.rules.values().filter(|r| r.default);
                 for rule in rules.filter(|r| r.kind == kind && (!r.host || host)) {
-                    self.fill(dep.name(rule.name), order, added);
+                    deps.push(self.build_graph(dep.name(rule.name), nodes, edges));
                 }
             } else {
-                self.fill(dep, order, added);
+                deps.push(self.build_graph(dep, nodes, edges));
             }
         }
-        order.push(step);
+
+        edges.entry(idx).or_insert(HashSet::new()).extend(deps);
+        return idx
+    }
+
+    /// Given a dependency graph with a finished list of `nodes`, fill out more
+    /// dependency `edges`.
+    ///
+    /// This is the step which satisfies all `after` listed dependencies in
+    /// `Rule` above.
+    fn satisfy_after_deps(&self,
+                          nodes: &HashMap<Step<'a>, usize>,
+                          edges: &mut HashMap<usize, HashSet<usize>>) {
+        // Reverse map from the name of a step to the node indices that it
+        // appears at.
+        let mut name_to_idx = HashMap::new();
+        for (step, &idx) in nodes {
+            name_to_idx.entry(step.name).or_insert(Vec::new()).push(idx);
+        }
+
+        for (step, idx) in nodes {
+            if *step == Step::noop() {
+                continue
+            }
+            for after in self.rules[step.name].after.iter() {
+                // This is the critical piece of an `after` dependency. If the
+                // dependency isn't actually in our graph then no edge is drawn,
+                // only if it's already present do we draw the edges.
+                if let Some(idxs) = name_to_idx.get(after) {
+                    edges.get_mut(idx).unwrap()
+                         .extend(idxs.iter().cloned());
+                }
+            }
+        }
+    }
+
+    fn topo_sort(&self,
+                 cur: usize,
+                 nodes: &HashMap<usize, &Step<'a>>,
+                 edges: &HashMap<usize, HashSet<usize>>,
+                 visited: &mut HashSet<usize>,
+                 order: &mut Vec<Step<'a>>) {
+        if !visited.insert(cur) {
+            return
+        }
+        for dep in edges[&cur].iter() {
+            self.topo_sort(*dep, nodes, edges, visited, order);
+        }
+        order.push(nodes[&cur].clone());
     }
 }
 

